cdc/log.cc

```diff
@@ -675,15 +675,102 @@ utils::UUID generate_timeuuid(api::timestamp_type t) {
     return utils::UUID_gen::get_random_time_UUID_from_micros(t);
 }
 
+class log_mutation_builder {
+    const schema& _base_schema;
+    const schema& _log_schema;
+    const column_definition& _op_col;
+    const column_definition& _ttl_col;
+
+    // The base mutation's partition key
+    const partition_key& _base_pk;
+
+    // The cdc$time value of created rows
+    const bytes _tuuid;
+    // The timestamp of the created log mutation cells
+    const api::timestamp_type _ts;
+    // The ttl of the created log mutation cells
+    const ttl_opt _ttl;
+
+    // Keeps the next cdc$batch_seq_no value
+    int _batch_no = 0;
+
+    // The mutation under construction
+    mutation& _log_mut;
+
+public:
+    log_mutation_builder(mutation& log_mut, api::timestamp_type ts,
+                         const partition_key& base_pk, const schema& base_schema)
+        : _base_schema(base_schema), _log_schema(*log_mut.schema()),
+          _op_col(*_log_schema.get_column_definition(log_meta_column_name_bytes("operation"))),
+          _ttl_col(*_log_schema.get_column_definition(log_meta_column_name_bytes("ttl"))),
+          _base_pk(base_pk),
+          _tuuid(timeuuid_type->decompose(generate_timeuuid(ts))),
+          _ts(ts),
+          _ttl(_base_schema.cdc_options().ttl()
+                  ? std::optional{std::chrono::seconds(_base_schema.cdc_options().ttl())} : std::nullopt),
+          _log_mut(log_mut)
+    {}
+
+    clustering_key allocate_new_log_row() {
+        const auto log_ck = clustering_key::from_exploded(_log_schema, { _tuuid, int32_type->decompose(_batch_no++) });
+        auto pk_value = _base_pk.explode(_base_schema);
```

Comment:
Would it make sense to store pk_value as a class field instead of _base_pk?

Reply:
It would